src/types/connection/stream.rs

```diff
@@ -0,0 +1,144 @@
+use crate::types::connection::{EmptyEdgeFields, QueryOperation};
+use crate::{Connection, Context, Cursor, DataSource, FieldResult, ObjectType};
+use futures::{Stream, StreamExt};
+use std::pin::Pin;
+
+pub struct StreamDataSource<'a, T, E = EmptyEdgeFields>
+where
+    T: Send + Sync + 'a,
+    E: ObjectType + Send + Sync + 'a,
+{
+    stream: Pin<Box<dyn Stream<Item = (Cursor, E, T)> + Send + Sync + 'a>>,
+}
+
+impl<'a, T, E> StreamDataSource<'a, T, E>
+where
+    T: Send + Sync + 'a,
+    E: ObjectType + Send + Sync + 'a,
+{
+    pub fn new<S>(stream: S) -> Self
+    where
+        S: Stream<Item = (Cursor, E, T)> + Send + Sync + 'a,
+    {
+        StreamDataSource {
+            stream: Box::pin(stream),
+        }
+    }
+}
```

Comment:
I've removed the mapping function for cursor, `StreamDataSource` now takes a `Stream<Item = (Cursor, E, T)>`. This gives more flexibility to users and is easy enouth for anyone to handle it in their codebase. Also gives the ability to control the EdgeFieldsObj. (see test)

Reply:
I think we should adopt the same strategy for the slice implementation.

Reply:
I agree. 😁

Reply:
removing the closure, we can now have `impl From<Stream> for StreamDataSource`, which is nice :)